File: firmware/cam-k230/hearth_mqtt.py
# hearth_mqtt — MQTT over umqtt.simple speaking FABRIC.md topics/payloads.
# birth/LWT (§2), announce (§3), event (§4), cmd/* dispatch (§7).
from umqtt.simple import MQTTClient
import ujson

class Mqtt:
    def __init__(self, host, port, device_id, kind, name, fw):
        self.device_id = device_id
        self.kind = kind
        self.name = name
        self.fw = fw
        self.base = "hearth/%s/" % device_id
        self.location = ""
        self.on_command = None
        self._announce = None
        will = (self.base + "status").encode()
        will_msg = ujson.dumps({"device_id": device_id, "kind": kind,
                                "online": False, "ts": 0}).encode()
        self.cli = MQTTClient(device_id, host, port=port, keepalive=20)
        self.cli.set_last_will(will, will_msg, retain=True, qos=0)
        self.cli.set_callback(self._cb)

    def _cb(self, topic, msg):
        t = topic.decode()
        i = t.find("/cmd/")
        if i >= 0 and self.on_command:
            self.on_command(t[i + 5:], msg.decode())

    def connect(self):
        self.cli.connect()
        self._birth()
        self.cli.subscribe((self.base + "cmd/#").encode())
        if self._announce:
            self.publish_announce(*self._announce)
        print("[mqtt] connected", self.device_id)

    def loop(self):
        try:
            self.cli.check_msg()
        except OSError:
            try: self.connect()
            except Exception as e: print("[mqtt] reconnect failed", e)

    def _birth(self):
        self._pub("status", {"device_id": self.device_id, "kind": self.kind,
                             "name": self.name, "online": True,
                             "fw": self.fw, "ts": 0}, retain=True)

    def _pub(self, suffix, obj, retain=False):
        self.cli.publish((self.base + suffix).encode(), ujson.dumps(obj).encode(), retain=retain)

    def publish_announce(self, endpoint, caps, transport="mqtt", instruments=None):
        self._announce = (endpoint, caps, transport, instruments)
        obj = {"device_id": self.device_id, "kind": self.kind, "name": self.name,
               "location": self.location, "fw": self.fw, "endpoint": endpoint,
               "transport": transport, "capabilities": caps}
        if instruments is not None:
            obj["instruments"] = instruments
        self._pub("announce", obj)

    def publish_event(self, ev_kind, confidence, thumb_b64, clip_url, ts):
        self._pub("event", {"device_id": self.device_id, "kind": ev_kind,
                            "confidence": round(confidence, 2),
                            "thumb_b64": thumb_b64, "clip_url": clip_url, "ts": ts})

File: firmware/common/hearth_mqtt.cpp
#include "hearth_mqtt.h"

namespace hearth {

// PubSubClient's callback is a C function pointer, so we route through a single
// instance. Only one Mqtt per firmware in practice (one hub link per device).
static Mqtt* s_self = nullptr;

void Mqtt::begin(const char* host, uint16_t port,
                 const String& deviceId, const char* kind,
                 const String& name, const String& fw) {
    s_self    = this;
    host_     = host;
    port_     = port;
    deviceId_ = deviceId;
    kind_     = kind;
    name_     = name;
    fw_       = fw;
    base_     = String("hearth/") + deviceId + "/";

    mqtt_.setServer(host_.c_str(), port_);
    mqtt_.setBufferSize(2048);            // events carry small base64 thumbnails
    mqtt_.setKeepAlive(20);
    mqtt_.setCallback(&Mqtt::trampoline);
}

void Mqtt::trampoline(char* topic, uint8_t* payload, unsigned len) {
    if (!s_self || !s_self->cmd_) return;
    String t(topic);
    String body; body.reserve(len);
    for (unsigned i = 0; i < len; ++i) body += (char)payload[i];
    // topic = hearth/<id>/cmd/<name>  -> extract <name>
    int cmdPos = t.indexOf("/cmd/");
    if (cmdPos < 0) return;
    s_self->cmd_(t.substring(cmdPos + 5), body);
}

bool Mqtt::connected() { return mqtt_.connected(); }

void Mqtt::publishBirth() {
    // §2 retained presence.
    String j = "{";
    j += "\"device_id\":\"" + deviceId_ + "\",";
    j += "\"kind\":\""      + String(kind_) + "\",";
    j += "\"name\":\""      + name_ + "\",";
    j += "\"online\":true,";
    j += "\"fw\":\""        + fw_ + "\",";
    j += "\"ts\":0}";
    publish("status", j, /*retained=*/true);
}

bool Mqtt::reconnect() {
    // LWT: same topic, online=false, retained — ungraceful drop marks offline.
    String willTopic = base_ + "status";
    String willMsg   = String("{\"device_id\":\"") + deviceId_ +
                       "\",\"kind\":\"" + kind_ + "\",\"online\":false,\"ts\":0}";
    bool ok = mqtt_.connect(deviceId_.c_str(), nullptr, nullptr,
                            willTopic.c_str(), 0, /*retained=*/true,
                            willMsg.c_str(), /*cleanSession=*/true);
    if (!ok) return false;

    publishBirth();
    if (haveAnnounce_)
        publishAnnounce(anEndpoint_, anTransport_, anCaps_, anInstr_);

    // Subscribe to all hub->device commands (§7).
    String sub = base_ + "cmd/#";
    mqtt_.subscribe(sub.c_str());
    Serial.printf("[mqtt] connected %s\n", deviceId_.c_str());
    return true;
}

void Mqtt::loop() {
    if (!mqtt_.connected()) {
        uint32_t now = millis();
        if (now - lastTry_ < 3000) return;   // backoff
        lastTry_ = now;
        reconnect();
        return;
    }
    mqtt_.loop();
}

bool Mqtt::publish(const char* suffix, const String& payload, bool retained) {
    if (!mqtt_.connected()) return false;
    String t = base_ + suffix;
    return mqtt_.publish(t.c_str(), (const uint8_t*)payload.c_str(),
                         payload.length(), retained);
}

void Mqtt::publishAnnounce(const String& endpoint, const char* transport,
                           const String& capsJson, const String& instrumentsJson) {
    // Cache for auto re-announce on reconnect.
    haveAnnounce_ = true;
    anEndpoint_   = endpoint;
    anTransport_  = transport;
    anCaps_       = capsJson;
    anInstr_      = instrumentsJson;

    String j = "{";
    j += "\"device_id\":\"" + deviceId_ + "\",";
    j += "\"kind\":\""      + String(kind_) + "\",";
    j += "\"name\":\""      + name_ + "\",";
    j += "\"location\":\""  + loc_ + "\",";
    j += "\"fw\":\""        + fw_ + "\",";
    j += "\"endpoint\":\""  + endpoint + "\",";
    j += "\"transport\":\"" + String(transport) + "\",";
    j += "\"capabilities\":" + capsJson;
    if (instrumentsJson.length())
        j += ",\"instruments\":" + instrumentsJson;
    j += "}";
    publish("announce", j, /*retained=*/false);
}

void Mqtt::publishEvent(const char* evKind, float confidence,
                        const String& thumbB64, const String& clipUrl, long ts) {
    String j = "{";
    j += "\"device_id\":\"" + deviceId_ + "\",";
    j += "\"kind\":\""      + String(evKind) + "\",";
    j += "\"confidence\":"  + String(confidence, 2) + ",";
    j += "\"thumb_b64\":\"" + thumbB64 + "\",";
    j += "\"clip_url\":\""  + clipUrl + "\",";
    j += "\"ts\":"          + String(ts) + "}";
    publish("event", j, /*retained=*/false);
}

void Mqtt::publishReading(const String& instrumentId, double value,
                          const char* unit, const char* deviceClass, long ts) {
    String j = "{";
    j += "\"instrument_id\":\"" + instrumentId + "\",";
    j += "\"device_id\":\""     + deviceId_ + "\",";
    j += "\"value\":"           + String(value, 4) + ",";
    j += "\"unit\":\""          + String(unit) + "\",";
    j += "\"device_class\":\""  + String(deviceClass) + "\",";
    j += "\"ts\":"              + String(ts) + "}";
    String suffix = String("reading/") + instrumentId;
    publish(suffix.c_str(), j, /*retained=*/true);
}

void Mqtt::publishWake(const String& phrase, long ts) {
    String j = String("{\"phrase\":\"") + phrase + "\",\"ts\":" + String(ts) + "}";
    publish("wake", j, /*retained=*/false);
}

} // namespace hearth
